snooty/postprocess.py

```diff
@@ -615,6 +616,59 @@ def add_titles_to_label_targets(
                     target.children = heading.children
             self.pending_targets = []
 
+    def bound_included_AST(
+        self,
+        nodes: MutableSequence[n.Node],
+        start_after_text: str,
+        end_before_text: str,
+    ) -> Tuple[MutableSequence[n.Node], bool, bool]:
+        """Given an AST in the form of nodes, return a subgraph of that AST by removing nodes 'outside' of 
+        the bound formed by the nodes containing the start_after_text or end_before_text. In in-order traversal,
+        a node is considered 'outside' the subgraph if it precedes and is not any ancestor of the start-after node,
+        or if it succeeds and is not any ancestor of the end-before node."""
+
+        def is_bound(node: n.Node, search_text: str) -> bool:
+            """Helper function to determine if the given node contains specified start-after or end-before text. 
+            
+            Note: For now, we are only splicing included files based on Text and TargetIdentifier nodes. 
+            Comments have Text nodes as children; Labels have TargetIdentifiers as children. 
+            So, writers can splice on these parent directives as well."""
+            if isinstance(node, n.Text) and node.value:
+                return search_text == node.value
+            elif isinstance(node, n.TargetIdentifier):
+                if node.ids and isinstance(node.ids[0], str):
+                    return search_text in node.ids
+            return False
+
+        start_index, end_index = 0, len(nodes)
+        any_start, any_end = False, False
+
+        # For any given node: if the start_after node is within this node's subtree, do not include any
+        # preceding siblings of this node in the resulting AST; if the end_before node is within this
+        # node's subtree, then do not include any succeeding siblings of this node.
+        for i, node in enumerate(nodes):
+            has_start, has_end = False, False
+            # Determine if this node itself (not a child node) contains a bound
+            is_start = is_bound(node, start_after_text)
+            is_end = is_bound(node, end_before_text)
+            # Recursively search the child nodes for bounds
+            if isinstance(node, n.Parent):
+                children, has_start, has_end = self.bound_included_AST(
+                    node.children, start_after_text, end_before_text
+                )
+                node.children = children
+            if is_start or has_start:
+                any_start = True
+                start_index = i
+            if is_end or has_end:
+                any_end = True
+                end_index = i
+        assert (
```

Comment:
The assert statement isn't safe to use in this kind of context: it is removed by the compiler in optimization mode. Assertions are to catch cases where the *developer* (you or me) has made a mistake, not where the user has provided something unexpected

Reply:
TIL! Thanks :)

Reply:
To be clear, are we invoking `-O` to remove asserts at runtime? Because when I put an `assert False` in `postprocess.py`, it catches when I build. Is this a continuous integration thing? (Also if `-O` is passed somewhere, can you point me to where it is?)

Reply:
We don't run with `-O`, but correct software needs to.

This also just isn't what assertions are for.